cartridge/upload.lua

```diff
@@ -0,0 +1,313 @@
+    checks('string', 'string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = vars.get_upload_path(upload_id)
+    local payload_path = fio.pathjoin(upload_path, 'payload')
+    local ok, err = utils.file_write(payload_path, payload)
+    fiber.testcancel()
+    if not ok then
+        return nil, err
+    end
+
+    vars.upload_fibers[upload_id] = nil
+
+    return true
+end
+
+-- The uploaded data is read from the shared resource on every instance,
+-- not only on the transmitter.
+local function upload_finish(upload_id)
+    checks('string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = vars.get_upload_path(upload_id)
+    local payload_path = fio.pathjoin(upload_path, 'payload')
+    local payload, err = utils.file_read(payload_path)
+    fiber.testcancel()
+    if err then
+        return nil, err
+    end
+
+    vars.upload_fibers[upload_id] = nil
+
+    local ok, data = pcall(msgpack.decode, payload)
+    if not ok then
+        return nil, UploadError:new(data)
+    end
+    vars.inbox[upload_id] = data
+    return true
+end
+
+-- Remove shared resources. Be idempotent.
+local function upload_cleanup(upload_id)
+    checks('string')
+    if vars.upload_fibers[upload_id] ~= nil then
+        pcall(fiber.cancel, vars.upload_fibers[upload_id])
+        vars.upload_fibers[upload_id] = nil
+    end
+
+    local upload_path = vars.get_upload_path(upload_id)
+    local random_path = utils.randomize_path(upload_path)
+
+    local ok = fio.rename(upload_path, random_path)
+    local _errno = errno()
+    if not ok then
+        if not fio.path.is_dir(upload_path) then
+            -- No such file, upload_path is already clean.
+            return true
+        else
+            log.warn(
+                'Error removing %s: %s',
+                upload_path, errno.strerror(_errno)
+            )
+            return false
+        end
+    end
+
+    local ok = fio.rmtree(random_path)
+    local _errno = errno()
+    if not ok then
+        log.warn(
+            "Error removing %s: %s",
+            random_path, errno.strerror(_errno)
+        )
+        return false
+    end
+
+    return true
+end
+
+--- Spread the data across the cluster.
+--
+-- For each separate upload, a random `upload_id` is generated. All the
+-- instances try to create `/tmp/<upload_id>` on their side, and those
+-- who succeed act as transmitters.
+--
+-- When the upload finishes, all the instances load the data into the
+-- `inbox` table and the temporary files are cleared. The inbox isn't
+-- garbage-collected automatically. It's the user's responsibility to
+-- clean it up after use.
+--
+-- @function upload
+--
+-- @param data
+--   any Lua object.
+-- @tparam {string,...} uri_list
+--   array of URIs.
+--
+-- @treturn[1] string `upload_id` (if at least one upload succeded)
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+local function upload(data, uri_list)
```

Comment:
This module is quite atomic and can be used by other projects, let's move timeout to options

Reply:
`upload_path ` also should be moved to options. I think all constants/options could be passed via arguments.
Cartridge should provide only default values and do it out of the scope this module.

Reply:
> This module is quite atomic and can be used by other projects, let's move timeout to options

Good idea. Done.



> `upload_path ` also should be moved to options. I think all constants/options could be passed via arguments.
> Cartridge should provide only default values and do it out of the scope this module.

And this seems to be impossible. The sending side can't know where the accepting side wants to store the data.